### PR TITLE
Time Travelling

### PR DESCRIPTION
**Pull request by glatteis**
Hi :)

This pull requests adds time travelling.

This pull request shouldn't change any default behaviour in storm. It adds two CLI flags:

- --time-travel to enable time travelling as a preprocessing
- --linear-to-simple to convert linear pMCs to simple pMCs which allows time travelling to be performed on them

There is a test for time travelling but currently no test for the linear-to-simple converter.

### CODE REVIEW

File: src/storm-pars-cli/storm-pars.cpp [hunk 1]
```diff
@@ -295,6 +300,25 @@ namespace storm {
...
+            
+            if (parametricSettings.isLinearToSimpleEnabled()) {
+                STORM_LOG_INFO("Transforming linear to simple...");
+                transformer::BinaryDtmcTransformer transformer;
+                result.model = transformer.transform(*result.model->template as<storm::models::sparse::Dtmc<RationalFunction>>(), true);
+                result.changed = true;
+            }
+
+            if (parametricSettings.isTimeTravellingEnabled()) {
+                // storm::transformer::BinaryDtmcTransformer transformer;
```
Comment (sjunges) [inline src/storm-pars-cli/storm-pars.cpp:312]: comments necessary?

File: src/storm-pars-cli/storm-pars.cpp [hunk 2]
```diff
@@ -48,6 +52,7 @@
 #include "storm/settings/modules/IOSettings.h"
 #include "storm/settings/modules/BisimulationSettings.h"
 #include "storm/settings/modules/TransformationSettings.h"
+#include "utility/logging.h"
```
Comment (sjunges) [inline src/storm-pars-cli/storm-pars.cpp:55]: this include is off from our typical setup

File: src/storm-pars/transformer/TimeTravelling.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
+#ifndef STORM_EQUALPARAMATERREDUCER_H
```
Comment (sjunges) [inline src/storm-pars/transformer/TimeTravelling.h:1]: Use pragma once (or make sure that the name matches the filename)

File: src/storm-pars/transformer/TimeTravelling.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     * @param matrix FlexibleMatrix of the pMC.
+     * @param alreadyVisited Recursive argument, set this to the empty map.
+     * @param treeStates The tree states (see updateTreeStates).
+     * @param allParameters The set of all parameters of the pMC.
+     * @param stateRewardVector The state reward vector of the pMC.
+     * @param stateLabelling The state labeling of the pMC.
+     * @param labelsInFormula The labels in the formula.
+     * @return false (returns true in recursive cases)
+     */
+    bool jipConvert(uint_fast64_t state, storage::FlexibleSparseMatrix<RationalFunction>& matrix, std::map<uint_fast64_t, bool>& alreadyVisited,
```
Comment (sjunges) [inline src/storm-pars/transformer/TimeTravelling.h:88]: Can we give thsi a more descriptive name?

File: src/storm-pars/transformer/BinaryDtmcTransformer.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "storm/models/sparse/Dtmc.h"
+#include "storm/models/sparse/StandardRewardModel.h"
+
+namespace storm {
+    namespace transformer {
+
+        class BinaryDtmcTransformer {
+
```
Comment (sjunges) [inline src/storm-pars/transformer/BinaryDtmcTransformer.h:10]: What are the preconditions for this dtmc transformer?

File: src/storm-pars/transformer/TimeTravelling.h [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+namespace storm {
+namespace transformer {
+
+class TimeTravelling {
+   public:
+    /**
+     * This class re-orders parameteric transitions of a pMC so bounds computed by Parameter Lifting will eventually be better.
+     * The parametric reachability probability for the given check task will be the same in the time-travelled and in the original model.
+     */
+    TimeTravelling() {}
```
Comment (sjunges) [inline src/storm-pars/transformer/TimeTravelling.h:22]: = default?

File: src/storm-pars/transformer/BinaryDtmcTransformer.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        public:
+            
+            BinaryDtmcTransformer();
```
Comment (sjunges) [inline src/storm-pars/transformer/BinaryDtmcTransformer.h:13]: = default()

**Comment by glatteis** [comment]
Thanks for the code review Sebastian! I'm here super late but I will get to this in the next month or so :)

**Comment by sjunges** [comment]
@glatteis Are you still planning to integrate these fixes? :-)

File: src/storm-pars/transformer/BinaryDtmcTransformer.h (hunk 5 above)
Comment (glatteis) [reply src/storm-pars/transformer/BinaryDtmcTransformer.h:10]: made a docstring :)

**Comment by glatteis** [comment]
let's see if this passes

**Comment by sjunges** [comment]
Hi,

Many thanks!
This looks mostly good to me. Would you still be able to add a test case for the linear-to-simple transformation? That would be really helpful!

Best,
Sebastian

File: src/storm-pars/transformer/TimeTravelling.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+     * @param stateRewardVector The state reward vector of the pMC.
+     * @param stateLabelling The state labeling of the pMC.
+     * @param labelsInFormula The labels in the formula.
+     * @return false (returns true in recursive cases)
+     */
+    bool collapseConstantTransitions(uint_fast64_t state, storage::FlexibleSparseMatrix<RationalFunction>& matrix, std::map<uint_fast64_t, bool>& alreadyVisited,
+                    const std::map<RationalFunctionVariable, std::map<uint_fast64_t, std::set<uint_fast64_t>>>& treeStates,
+                    const std::set<carl::Variable>& allParameters, const boost::optional<std::vector<RationalFunction>>& stateRewardVector,
+                    const models::sparse::StateLabeling stateLabelling, const std::set<std::string> labelsInFormula);
+    bool labelsIntersectedEqual(const std::set<std::string>& labels1, const std::set<std::string>& labels2, const std::set<std::string>& intersection);
```
Comment (sjunges) [inline src/storm-pars/transformer/TimeTravelling.h:91]: I would not make this utility function part of the class, but just define this in the cpp file

File: src/storm-pars/transformer/TimeTravelling.h [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+     * @param treeStates The tree states map to update.
+     * @param workingSets Where to start the search. When creating the tree states map: set this to all states with parametric transitions.
+     * @param flexibleMatrix The flexible matrix of the pMC.
+     * @param allParameters The set of all parameters of the pMC.
+     * @param stateRewardVector The state reward vector of the pMC.
+     * @param stateLabelling The state labelling of the pMC.
+     * @param labelsInFormula The labels that occur in the property.
+     */
+    void updateTreeStates(std::map<RationalFunctionVariable, std::map<uint_fast64_t, std::set<uint_fast64_t>>>& treeStates,
+                          std::map<RationalFunctionVariable, std::set<uint_fast64_t>>& workingSets,
```
Comment (sjunges) [inline src/storm-pars/transformer/TimeTravelling.h:47]: uint64_t vs uint_fast64_t  is inconsistent in this PR. We prefer uint64_t.

**Comment by glatteis** [comment]
I'm currently working on the test bit by bit :)

**Comment by glatteis** [comment]
working on it today

**Comment by glatteis** [comment]
I couldn't reproduce the issue but I think that was because the CI is pulling the target branch into this PR, which makes sense. I think the issue was probably the switch from `boost::optional` to `std::optional`, but I haven't build this locally yet

**Comment by glatteis** [comment]
Looks good! Sorry for taking so long

**Comment by glatteis** [comment]
Im closing and re-opening the PR to re-run the checks on the newer master branch

**Comment by sjunges** [comment]
LGTM, @tquatmann shall we squash and merge?

**Pull request merged**